In [1]:
# these warnings are fine. you can ignore them.
from util import *
import numpy as np
import random, math

device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
print("Imports finished.")

/home/niyati/miniconda3/lib/python3.11/site-packages/torch_geometric/typing.py:54: UserWarning: An issue occurred while importing 'pyg-lib'. Disabling its usage. Stacktrace: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/niyati/miniconda3/lib/python3.11/site-packages/libpyg.so)
  warnings.warn(f"An issue occurred while importing 'pyg-lib'. "
/home/niyati/miniconda3/lib/python3.11/site-packages/torch_geometric/typing.py:110: UserWarning: An issue occurred while importing 'torch-sparse'. Disabling its usage. Stacktrace: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by /home/niyati/miniconda3/lib/python3.11/site-packages/libpyg.so)
  warnings.warn(f"An issue occurred while importing 'torch-sparse'. "


Imports finished.


# Setting Up Dataset/Model/Ground Truth

In [2]:
dataset = Dataset(root='/tmp/Cora', name='Cora', device=device)
data, in_feats, h_feats, num_classes = dataset.get_data()

model = get_model(in_feats, h_feats, num_classes)

ground_truth = get_ground_truth(model, data)
print(ground_truth)

0.749


# Experiments
**Note:** The ground truth here is $0.749$

**Note:** This experiment was done before the development of new functions. The experiments will not work until transferred over to current function.

#### Accidental Degree Experiment

In [12]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 100

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0020
Change in edges:  2739.0  | Percentage change: 103.79%


In [9]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 200

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0010
Change in edges:  2839.0  | Percentage change: 107.58%


In [10]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 300

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0120
Change in edges:  2939.0  | Percentage change: 111.37%


In [13]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 400

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0130
Change in edges:  3039.0  | Percentage change: 115.16%


In [14]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 500

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0190
Change in edges:  3139.0  | Percentage change: 118.95%


In [15]:
from collections import defaultdict
import torch_geometric.utils as pyg_utils
# Assuming reset_dataset() is defined and initializes the graph correctly.
data = dataset.get_data()[0]
G = pyg_utils.to_networkx(data, to_undirected=True)

init_edges = G.number_of_edges()

num_nodes = G.number_of_nodes()
num_edges_to_add = 600

#existing_edges = set(modified_graph.edges())
potential_edges = [(i, j) for i in range(num_nodes) for j in range(i + 1, num_nodes) if not G.has_edge(i, j)]

random.shuffle(potential_edges)  # Shuffle to randomize which edges are considered
edges_to_add = potential_edges[:num_edges_to_add]

for i, j in edges_to_add:
    G.add_edge(i, j)



modified_graph = convert_to_pyg(G, x, y, train_mask, test_mask)
final_edges = len(modified_graph.edge_index[1])

output_accuracy_change(ground_truth, test_model(model, modified_graph)) 
number_added_edges(init_edges, final_edges, is_undirected=True)


----
The accuracy has changed by -0.0130
Change in edges:  3239.0  | Percentage change: 122.74%
